File: Advanced/TemplaterServer/src/main/java/hr/ngs/templater/server/LibreOffice.java
package hr.ngs.templater.server;

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.util.Date;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;

public class LibreOffice implements PdfConverter {
	private final Logger logger;
	private final int timeoutLimit;
	private final String tmpFolder;
	private final String libreOfficePath;
	private int counter = 1;

	public LibreOffice(Logger logger, String[] args) {
		this.logger = logger;
		int timeout = 30;
		String tmp = "";
		String path = "";
		for (String a : args) {
			if (a.startsWith("-timeout=")) {
				timeout = Integer.parseInt(a.substring("-timeout=".length()));
			} else if (a.startsWith("-tmp=")) {
				tmp = a.substring("-tmp=".length());
				File f = new File(tmp);
				if (!f.exists()) {
					throw new RuntimeException("Unable to find specified temporary folder: " + tmp);
				}
			} else if (a.startsWith("-libreoffice=")) {
				path = a.substring("-libreoffice=".length());
			}
		}
		if (path.length() == 0) {
			File pf = new File("C:/Program Files (x86)");
			if (pf.exists()) {
				File[] files = pf.listFiles(f -> f.getName().startsWith("LibreOffice "));
				if (files != null && files.length > 0) {
					File prog = new File(files[0], "program");
					File soffice = new File(prog, "soffice.exe");
					if (soffice.exists()) {
						path = soffice.getAbsolutePath();
					}
				}
			}
			if (path.length() == 0) {
				throw new RuntimeException(
						"Unable to find LibreOffice on the system. Please explicitly specify it via: -libreoffice=/user/home/office/libreoffice");
			}
		}
		this.timeoutLimit = timeout;
		this.tmpFolder = tmp;
		this.libreOfficePath = path;
	}

	@Override
	public synchronized byte[] convert(byte[] template, String extension) throws Exception {

			final File tmpFile = tmpFolder.length() == 0
					? File.createTempFile("templaterDocument", "." + extension)
					: new File(tmpFolder, "templaterDocument" + (counter++) + "." + extension);
			final String outputFileName = tmpFile.getPath().substring(0, tmpFile.getPath().length() - extension.length()) + "pdf";

			try {
				long start = new Date().getTime();
				final OutputStream os = new FileOutputStream(tmpFile);
				os.write(template);
				os.close();

				ProcessBuilder builder = new ProcessBuilder(libreOfficePath, "--norestore", "--nofirststartwizard", "--nologo", "--headless", "--convert-to", "pdf", tmpFile.getPath());
				builder.directory(tmpFile.getParentFile());
				Process process = builder.start();
				if (process.waitFor(timeoutLimit, TimeUnit.SECONDS)) {
					File result = new File(outputFileName);
					try {
						if (result.exists()) {
							byte[] output = Files.readAllBytes(result.toPath());
							if (logger.isLoggable(Level.FINE)) {
								logger.log(Level.FINE, String.format("PDF conversion finished in %d ms. input size = %d, output size = %d", new Date().getTime() - start, template.length, output.length));
							}
							return output;
						} else {
							if (logger.isLoggable(Level.WARNING)) {
								logger.log(Level.WARNING, String.format("Unable to find output PDF. Duration: %d ms", new Date().getTime() - start));
							}
						}
					} finally {
						result.delete();
					}
				} else {
					if (logger.isLoggable(Level.WARNING)) {
						logger.log(Level.WARNING, String.format("Timeout waiting for PDF conversion. Duration: %d ms", new Date().getTime() - start));
					}
				}
			} finally {
				tmpFile.delete();
			}
			return null;
		}

}

File: Advanced/TemplaterServer/src/LibreOffice.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Linq;

namespace TemplaterServer.src
{
	class LibreOffice : PdfConverter
	{
		private readonly string LibreOfficePath;
		private readonly int Timeout;

		public LibreOffice(string[] args)
		{
			int timeout = 30;
			var path = "";
			foreach(var a in args)
			{
				if (a.StartsWith("timeout="))
					timeout = int.Parse(a.Substring("timeout=".Length));
				else if (a.StartsWith("libreoffice="))
				{
					path = a.Substring("libreoffice=".Length);
				}
			}
			if (path.Length == 0)
			{
				var macFI = new FileInfo("/Applications/LibreOffice.app/Contents/MacOS/soffice");
				var linuxFI = new FileInfo("/usr/bin/libreoffice");
				var win32 = new DirectoryInfo("C:/Program Files (x86)");
				var win64 = new DirectoryInfo("C:/Program Files");
				if (macFI.Exists)
					path = macFI.ToString();
				else if (linuxFI.Exists)
					path = linuxFI.ToString();
				else 
					path = FindWindows(win32) ?? FindWindows(win64);
				if (string.IsNullOrEmpty(path))
					throw new ArgumentException("Unable to find LibreOffice on the system. Please explicitly specify it via: -libreoffice=C:/Program Files (x86)/LibreOffice 6.4/program/soffice.exe");
			}
			this.Timeout = timeout * 1000;
			this.LibreOfficePath = path;
		}
		
		private static string FindWindows(DirectoryInfo folder)
		{
			if (!folder.Exists) return null;
			var nested = folder.GetDirectories();
			if (nested == null || nested.Length == 0) return null;
			foreach (var lo in nested)
			{
				if (!lo.FullName.Contains("LibreOffice", StringComparison.OrdinalIgnoreCase)) continue;
				var file = lo.FullName.Replace('\\', '/') + "/program/soffice.exe";
				if (File.Exists(file))
					return file;
			}
			return null;
		}		

		private int FileCounter;

		public Stream Convert(Stream template, string extension)
		{
			lock (this)
			{
				var tmpPath = Path.GetTempPath();
				var fileName = "templateDocument" + (++FileCounter) + "." + extension;
				var tmpFile = Path.Combine(tmpPath, fileName);
				using (var fs = File.OpenWrite(tmpFile))
					template.CopyTo(fs);
				try
				{
					var info = new ProcessStartInfo
					{
						FileName = LibreOfficePath,
						Arguments = "--norestore --nofirststartwizard --nologo --headless --convert-to pdf " + fileName,
						WorkingDirectory = tmpPath
					};
					var conv = Process.Start(info);
					conv.Start();
					if (!conv.WaitForExit(Timeout))
						throw new TimeoutException("Timeout waiting for PDF conversion");
					var pdfFile = Path.Combine(tmpPath, Path.GetFileNameWithoutExtension(fileName) + ".pdf");
					var bytes = File.ReadAllBytes(pdfFile);
					File.Delete(pdfFile);
					return new MemoryStream(bytes);
				}
				finally
				{
					File.Delete(tmpFile);
				}
			}
		}
	}
}
